javatests/arcs/core/storage/CapabilitiesResolverTest.kt

```diff
@@ -0,0 +1,140 @@
+import arcs.jvm.storage.database.testutil.MockDatabaseFactory
+import com.google.common.truth.Truth.assertThat
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+import java.lang.Error
+
+/** Tests for [CapabilitiesResolver]. */
+@RunWith(JUnit4::class)
+class CapabilitiesResolverTest{
+    @Before
+    fun setUp() {
+        VolatileDriverProvider(ArcId.newForTest("test"))
+    }
+
+    @After
+    fun tearDown() {
+        CapabilitiesResolver.reset()
+        CapabilitiesResolver.defaultCreators.clear()
+    }
+
+    @Test
+    fun capabilitiesResolver_createsStorageKeysDefault() {
+        val options = CapabilitiesResolver.StorageKeyOptions(ArcId.newForTest("test"))
+        val resolver = CapabilitiesResolver(options)
+        assertThat(resolver.findStorageKeyProtocols(Capabilities.TiedToArc)).containsExactly("volatile")
+        assertThat(resolver.findStorageKeyProtocols(Capabilities.TiedToRuntime).isEmpty()).isTrue()
+        assertThat(resolver.findStorageKeyProtocols(Capabilities.Persistent).isEmpty()).isTrue()
+        assertThat(resolver.createStorageKey(Capabilities.TiedToArc) is VolatileStorageKey).isTrue()
+        assertThrows(IllegalStateException::class) {
+            resolver.createStorageKey(Capabilities.TiedToRuntime)
+        }
+        assertThrows(IllegalStateException::class) {
+            resolver.createStorageKey(Capabilities.Persistent)
+        }
+        assertThrows(IllegalStateException::class) {
+            resolver.createStorageKey(Capabilities(setOf(Capabilities.Capability.TiedToArc, Capabilities.Capability.Persistent)))
+        }
+    }
+
+    @Test
+    fun capabilitiesResolver_createsStorageKeysCtor() {
+        val options = CapabilitiesResolver.StorageKeyOptions(ArcId.newForTest("test"))
+        val resolver = CapabilitiesResolver(options,
+            mutableMapOf("ramdisk" to CapabilitiesResolver.CapabilitiesCreator(
+                Capabilities.TiedToRuntime, { (arcId), _ -> RamDiskStorageKey(arcId.toString()) }
+            ))
+        )
+        assertThrows(IllegalStateException::class) {
+            resolver.createStorageKey(Capabilities.TiedToArc)
+        }
+        assertThat(resolver.createStorageKey(Capabilities.TiedToRuntime) is RamDiskStorageKey).isTrue()
+    }
+
+    @Test
+    fun capabilitiesResolver_createsStorageKeys() {
+        RamDisk.clear()
+        DatabaseDriverProvider.configure(MockDatabaseFactory(), mapOf<String, Schema>()::get)
+        val options = CapabilitiesResolver.StorageKeyOptions(ArcId.newForTest("test"))
+        val resolver1 = CapabilitiesResolver(options)
+        assertThat(resolver1.findStorageKeyProtocols(Capabilities.TiedToArc)).containsExactly("volatile")
+        assertThat(resolver1.findStorageKeyProtocols(Capabilities.TiedToRuntime)).containsExactly("ramdisk")
+        assertThat(resolver1.findStorageKeyProtocols(Capabilities.Persistent)).containsExactly("db")
```

Comment:
Suggest using the constants for the protocol names rather than string literals here.

Reply:
done